File: rust/buffer_ops.rs
//! Deprecated buffer helpers retained for reference.
//! Ref: 299effb7-cba4-41dd-9bf2-ecd15ed69a82
//!
//! Not linked into any build target; no public API.

compile_error!(
    "buffer_ops is not usable; retained for migration reference. \
     Ref: 299effb7-cba4-41dd-9bf2-ecd15ed69a82"
);

#[cfg(any())]
mod _dead {
    use std::ptr;

    /* Shape 1: fixed-length copy from a compile-time byte literal.
     * `src` is a constant, `n` is its length; the const assertion pins it
     * to the destination size. */
    unsafe fn bufops_copy_banner(dst: *mut u8) {
        const BANNER: &[u8] = b"status: ok\0";
        const _: () = assert!(BANNER.len() <= 64, "banner must fit in dst");
        ptr::copy_nonoverlapping(BANNER.as_ptr(), dst, BANNER.len());
    }

    /* Shape 2: variable-length copy with clamp one line above.
     * `if n > dst_cap { n = dst_cap; }` bounds the write. */
    unsafe fn bufops_copy_bounded(
        dst: *mut u8,
        dst_cap: usize,
        src: *const u8,
        mut n: usize,
    ) -> usize {
        if n > dst_cap {
            n = dst_cap;
        }
        if n == 0 {
            return 0;
        }
        ptr::copy_nonoverlapping(src, dst, n);
        n
    }

    /* Shape 3: truncating copy with explicit NUL.
     * `n <= dst_cap - 1` so the copy hits `[0, dst_cap-2]`; the NUL write
     * at `dst.add(n)` hits at most `dst_cap - 1`, in-bounds.
     * `dst_cap == 0` early-return handles the degenerate case. */
    unsafe fn bufops_copy_truncating(
        dst: *mut u8,
        dst_cap: usize,
        src: *const u8,
        src_len: usize,
    ) {
        if dst_cap == 0 {
            return;
        }
        let n = if src_len < dst_cap - 1 {
            src_len
        } else {
            dst_cap - 1
        };
        ptr::copy_nonoverlapping(src, dst, n);
        *dst.add(n) = 0;
    }

    /* Shape 4: copy within a single buffer (overlapping regions).
     * `ptr::copy` explicitly supports overlap (memmove equivalent);
     * the two guards bound `i + n` and `j + n` to `cap`. */
    unsafe fn bufops_shift(
        buf: *mut u8,
        cap: usize,
        i: usize,
        j: usize,
        n: usize,
    ) {
        if i > cap || n > cap - i {
            return;
        }
        if j > cap || n > cap - j {
            return;
        }
        ptr::copy(buf.add(i), buf.add(j), n);
    }
}

File: go/buffer_ops.go
//go:build ignore

// Deprecated buffer helpers retained for reference.
// Ref: ae4499ae-9474-423a-9dee-26751f95ffb0
//
// Not linked into any build target; no exported API.

package legacy

import "unsafe"

func init() {
	panic(
		"buffer_ops is not usable; retained for migration reference. " +
			"Ref: ae4499ae-9474-423a-9dee-26751f95ffb0",
	)
}

/* Shape 1: fixed-length copy from a compile-time literal.
 * src is a constant, n is its length; the compile-time assertion
 * pins it to the destination size. */
func bufopsCopyBanner(dst *[64]byte) {
	const banner = "status: ok"
	_ = [1]struct{}{}[len(banner)-1-(64-1)*0] // compile-time: len(banner) <= 64
	copy(dst[:], banner)
}

/* Shape 2: variable-length copy with clamp one line above.
 * if n > dstCap { n = dstCap } bounds the write. */
func bufopsCopyBounded(dst []byte, src []byte, n int) int {
	dstCap := len(dst)
	if n > dstCap {
		n = dstCap
	}
	if n == 0 {
		return 0
	}
	srcSlice := unsafe.Slice(&src[0], n)
	dstSlice := unsafe.Slice(&dst[0], n)
	copy(dstSlice, srcSlice)
	return n
}

/* Shape 3: truncating copy with explicit NUL.
 * n <= dstCap - 1 so the copy hits [0, dstCap-2]; the NUL write
 * at dst[n] hits at most dstCap - 1, in-bounds.
 * dstCap == 0 early-return handles the degenerate case. */
func bufopsCopyTruncating(dst []byte, src []byte, srcLen int) {
	dstCap := len(dst)
	if dstCap == 0 {
		return
	}
	n := srcLen
	if n > dstCap-1 {
		n = dstCap - 1
	}
	copy(unsafe.Slice(&dst[0], n), unsafe.Slice(&src[0], n))
	dst[n] = 0
}

/* Shape 4: copy within a single buffer (overlapping regions).
 * copy() supports overlap; the two guards bound i+n and j+n to cap. */
func bufopsShift(buf []byte, i, j, n int) {
	cap := len(buf)
	if i > cap || n > cap-i {
		return
	}
	if j > cap || n > cap-j {
		return
	}
	copy(buf[j:j+n], buf[i:i+n])
}
